src/main/java/com/pusher/client/channel/impl/PrivateEncryptedChannelImpl.java

```diff
@@ -102,6 +105,81 @@ private void tearDownChannel() {
         }
     }
 
+    @Override
+    public PusherEvent prepareMessage(final String event, String message) {
+        try {
+            Map receivedMessage = GSON.fromJson(message, Map.class);
+            final String decryptedMessage = decryptMessage((String)receivedMessage.get("data"));
+            receivedMessage.replace("data", decryptedMessage);
+
+            return GSON.fromJson(
+                    GSON.toJson(receivedMessage), PusherEvent.class);
+        } catch (AuthorizationFailureException e1) {
+
+            //dispose of the "old" secretboxopener
+            tearDownChannel();
+            //get a new secretboxopener
+            String authentication = authenticate();
+            //try to decrypt it again... (we'll try to make this better later)
+            try {
+                Map receivedMessage = GSON.fromJson(message, Map.class);
+                final String decryptedMessage = decryptMessage((String)receivedMessage.get("data"));
+                receivedMessage.replace("data", decryptedMessage);
+
+                return GSON.fromJson(
+                        GSON.toJson(receivedMessage), PusherEvent.class);
+            } catch (AuthorizationFailureException e2) {
+                notifyListenersOfDecryptFailure(event);
+            }
+
+        }
+
+        return null;
+    }
+
+    private void notifyListenersOfDecryptFailure(final String event) {
+        final Set<SubscriptionEventListener> listeners;
+        synchronized (lock) {
+            final Set<SubscriptionEventListener> sharedListeners = eventNameToListenerMap.get(event);
+            if (sharedListeners != null) {
+                listeners = new HashSet<SubscriptionEventListener>(sharedListeners);
+            }
+            else {
+                listeners = null;
+            }
+        }
```

Comment:
could be extracted to the parent class as `getInterestedListeners(String event)`